### ISSUE TITLE
Port to python3?

### ISSUE THREAD

**Issue by trofi** [abathur/resholve#112]
Recently `nixpkgs` started using `resholve` in `xdg-utils package. I only noticed it due to unrelated failures: `python2` packages use some old `setuptools` version that don't support `--parallel` (I use `enableParallelBuildingByDefault = true;`). I was surprised to see that something still uses `python2` to build my desktop system.

Would it be reasonable to post `resholve` to `python3` ecosystem? Would it take a lot of effort? Or perhaps it's blocked on upstream depends stuck on `python2`?

**Comment by SuperSandro2000** [comment]
See #88

**Comment by abathur** [comment]
This (or something like it) should eventually be ripe, btw. The c++ version of OSH is very close to parity with the reference python implementation, at which point some work on a new way to drive the parser and get a parse tree wouls clear the way for the refactoring needed to make use of it. 

resholve lives right on top of the existing parser (I actually monkey-patch it to collect interesting nodes as they are instantiated to save us the time of a full walk) so there's some chance that could be a slog. To prepare for that, I've already done a fair amount of refactoring to resholve to mostly-contain the interactions with the parse tree in our own abstractions. 

Porting should either be easy enough in that process, or easy enough to tack on after it. (The actual version port shouldn't be much work here tbh. This isn't an _old_ py2 codebase. All of my other Python work is py3.)

**Comment by trofi** [comment]
For the time being I proposed https://github.com/NixOS/nixpkgs/pull/288407 for `nixpkgs` to make `python2` packages compatible with `enableParallelBuildingByDefault`.

**Comment by anund** [comment]
Looks like https://github.com/NixOS/nixpkgs/pull/288407 is blocked on lack of review. :(

**Comment by bengsparks** [comment]
@abathur can you provide an update on the progress of porting `resholve` to Python3 / removing Python2 code from the repository?
